src/index/txindex.cpp

```diff
@@ -0,0 +1,252 @@
+
+    // Attempt to migrate txindex from the old database to the new one. Even if
+    // chain_tip is null, the node could be reindexing and we still want to
+    // delete txindex records the old database.
+    if (!m_db->MigrateData(*pblocktree, tip_hash)) {
+        return false;
+    }
+
+    if (!chain_tip) {
+        m_synced = true;
+        return true;
+    }
+
+    uint256 best_block_hash;
+    if (!m_db->ReadBestBlockHash(best_block_hash)) {
+        FatalError("%s: Failed to read from tx index database", __func__);
+        return false;
+    }
+
+    if (best_block_hash.IsNull()) {
+        return true;
+    }
+
+    auto it = mapBlockIndex.find(best_block_hash);
+    if (it == mapBlockIndex.end()) {
+        FatalError("%s: Last block synced by txindex is unknown", __func__);
+        return false;
+    }
+
+    const auto pindex = it->second;
+    m_best_block_index = pindex;
+    if (pindex->GetAncestor(chain_tip->nHeight) == chain_tip) {
+        m_synced = true;
+    }
+
+    return true;
+}
+
+static const CBlockIndex* NextSyncBlock(const CBlockIndex* pindex_prev)
+{
+    AssertLockHeld(cs_main);
+
+    if (!pindex_prev) {
+        return chainActive.Genesis();
+    }
+
+    auto pindex = chainActive.Next(pindex_prev);
+    if (pindex) {
+        return pindex;
+    }
+
+    return chainActive.Next(chainActive.FindFork(pindex_prev));
+}
+
+void TxIndex::ThreadSync()
+{
+    auto pindex = m_best_block_index.load();
+    if (!m_synced) {
+        auto& consensus_params = Params().GetConsensus();
+
+        int64_t last_log_time = 0;
+        while (true) {
+            if (m_interrupt) {
+                return;
+            }
+
+            {
+                LOCK(cs_main);
+                auto pindex_next = NextSyncBlock(pindex);
+                if (!pindex_next) {
+                    m_best_block_index = pindex;
+                    m_synced = true;
+                    break;
+                }
+                pindex = pindex_next;
+            }
+
+            int64_t current_time = GetTime();
+            if (last_log_time + SYNC_LOG_INTERVAL < current_time) {
+                LogPrintf("Syncing txindex with block chain from height %d\n", pindex->nHeight);
+                last_log_time = current_time;
+            }
+
+            CBlock block;
+            if (!ReadBlockFromDisk(block, pindex, consensus_params)) {
+                FatalError("%s: Failed to read block %s from disk",
+                           __func__, pindex->GetBlockHash().ToString());
+                return;
+            }
+            if (!WriteBlock(block, pindex)) {
+                FatalError("%s: Failed to write block %s to tx index database",
+                           __func__, pindex->GetBlockHash().ToString());
+                return;
+            }
+        }
+    }
+
+    if (pindex) {
+        LogPrintf("txindex is enabled at height %d\n", pindex->nHeight);
+    } else {
+        LogPrintf("txindex is enabled\n", pindex->nHeight);
```

Comment:
This is wrong, you have an extra argument that dereferences a nullptr.

Reply:
Thanks.